plugins/auth-backend/src/database/OidcDatabase.ts

```diff
@@ -0,0 +1,379 @@
+          throw new Error(
+            `Failed to retrieve updated authorization code with code ${authorizationCode.code}`,
+          );
+        }
+
+        return this.rowToAuthorizationCode(updated) as AuthorizationCode;
+      });
+    }
+
+    const [updated] = await this.db<OidcAuthorizationCodeRow>(
+      'oidc_authorization_codes',
+    )
+      .where('code', authorizationCode.code)
+      .update(updatedFields)
+      .returning('*');
+
+    return this.rowToAuthorizationCode(updated) as AuthorizationCode;
+  }
+
+  private rowToClient(row: Partial<OidcClientRow>): Partial<Client> {
+    return {
+      clientId: row.client_id,
+      clientName: row.client_name,
+      clientSecret: row.client_secret,
+      redirectUris: row.redirect_uris
+        ? JSON.parse(row.redirect_uris)
+        : undefined,
+      responseTypes: row.response_types
+        ? JSON.parse(row.response_types)
+        : undefined,
+      grantTypes: row.grant_types ? JSON.parse(row.grant_types) : undefined,
+      scope: row.scope ?? undefined,
+      metadata: row.metadata ? JSON.parse(row.metadata) : undefined,
+    };
+  }
+
+  private authorizationSessionToRow(
+    session: Partial<AuthorizationSession>,
+  ): Partial<OAuthAuthorizationSessionRow> {
+    return {
+      id: session.id,
+      client_id: session.clientId,
+      user_entity_ref: session.userEntityRef,
+      redirect_uri: session.redirectUri,
+      scope: session.scope,
+      state: session.state,
+      response_type: session.responseType,
+      code_challenge: session.codeChallenge,
+      code_challenge_method: session.codeChallengeMethod,
+      nonce: session.nonce,
+      status: session.status,
+      expires_at: session.expiresAt,
+    };
+  }
+
+  private rowToAuthorizationSession(
+    row: Partial<OAuthAuthorizationSessionRow>,
+  ): Partial<AuthorizationSession> {
+    return {
+      id: row.id,
+      clientId: row.client_id,
+      userEntityRef: row.user_entity_ref ?? undefined,
+      redirectUri: row.redirect_uri,
+      scope: row.scope ?? undefined,
+      state: row.state ?? undefined,
+      responseType: row.response_type,
+      codeChallenge: row.code_challenge ?? undefined,
+      codeChallengeMethod: row.code_challenge_method ?? undefined,
+      nonce: row.nonce ?? undefined,
+      status: row.status,
+      expiresAt: toDate(row.expires_at),
+    };
+  }
+
+  private authorizationCodeToRow(
+    authorizationCode: Partial<AuthorizationCode>,
+  ): Partial<OidcAuthorizationCodeRow> {
+    return {
+      code: authorizationCode.code,
+      session_id: authorizationCode.sessionId,
+      expires_at: authorizationCode.expiresAt,
+      used: authorizationCode.used,
+    };
+  }
+
+  private rowToAuthorizationCode(
+    row: Partial<OidcAuthorizationCodeRow>,
+  ): Partial<AuthorizationCode> {
+    return {
+      code: row.code,
+      sessionId: row.session_id,
+      expiresAt: toDate(row.expires_at),
+      used: Boolean(row.used),
```

Comment:
i'm a bit torn on these methods accepting a Partial input, it makes it so easy to forget like here that you need to have a `row.used === undefined ? Boolean(row.used) : undefined` in a ton of places. Could it not be made in such a way that the input never is a Partial? would sure simplify _some_ things at least. I think you may have missed some above but not sure

Reply:
Cleaned up what I could :pray: